### PR TITLE
[DRAFT] KOKORO-82M Model Support on Blackhole P150

### PR DESCRIPTION
**Pull request by ign-ssunil**
## Summary

Adds a TTNN implementation of **Kokoro-82M** (`hexgrad/Kokoro-82M`) — an 82M-parameter open-weight
**StyleTTS2 / ISTFTNet** text-to-speech model that generates 24 kHz speech from text, ported to
Tenstorrent Blackhole.

### This PR includes
- Implements the full on-device pipeline under `models/experimental/kokoro/` (one TTNN class per
  module in `tt/`, CPU reference in `reference/`):
  - **PL-BERT backbone** (`TTCustomAlbert`) — 12-layer contextual token encoder
  - **Prosody predictor** (`TTProsodyPredictor`) — duration BiLSTM → `pred_dur`; `F0Ntrain` → F0
    (pitch) + N (energy)
  - **ASR text encoder** (`TTTextEncoder`) — CNN + BiLSTM → aligned `asr` features
  - **ISTFTNet decoder/vocoder** (`TTDecoder` / `TTGenerator`) — AdaIN upsampling generator +
    harmonic-plus-noise source (`SineGen` → `SourceModuleHnNSF`) + STFT/iSTFT → 24 kHz waveform
- Adds `TTKModel` (top-level) with a metal-trace path (Trace A: prosody+ASR / host alignment step /
  Trace B: decoder) and a demo CLI (`demo/ttnn_kokoro_full_demo.py`) for text-to-WAV
- Adds per-module PCC tests, full-pipeline PCC (fallback matrix), quality metrics (Whisper WER /
  mel-PCC / speaker cosine / cFW2VD), and metal-trace capture/replay tests — all at full HF config
  with real Kokoro-82M weights
- Two opt-in CPU float32 fallbacks (SineGen phase, STFT) for maximum numerical parity; default is
  fully on-device

## Supported HW

| Device | Mesh |
|--------|------|
| **Blackhole (P150-class)** | 1×1 (single device) |

### Accuracy (full-pipeline audio PCC vs CPU reference)
```bash
# recommended config E (STFT + phase CPU fallbacks)
pytest models/experimental/kokoro/tests/test_tt_kmodel_pcc.py \
  -k test_tt_kmodel_stft_and_phase_fallback_pcc -sv --timeout=600

# no-fallback (shows the BH BF16 harmonic-source ceiling)
pytest models/experimental/kokoro/tests/test_tt_kmodel_pcc.py \
  -k test_tt_kmodel_generator_no_torch_fallback_pcc -sv --timeout=600
```

### Quality (WER, mel-PCC, speaker cosine, cFW2VD)
```bash
pytest models/experimental/kokoro/tests/test_tt_kmodel_asr_wer.py        -sv --timeout=3600
pytest models/experimental/kokoro/tests/test_tt_kmodel_mel_pcc.py        -sv --timeout=3600
pytest models/experimental/kokoro/tests/test_tt_kmodel_speaker_cosine.py -sv --timeout=3600
pytest models/experimental/kokoro/tests/test_tt_kmodel_cfw2vd.py         -sv --timeout=3600
```

### Performance
The demo reports per-chunk latency, RTF, time-to-first-audio, and char/s (metal trace on by default):
```bash
python models/experimental/kokoro/demo/ttnn_kokoro_full_demo.py --text "..." --voice af_heart
```

### Demo
```bash
python models/experimental/kokoro/demo/ttnn_kokoro_full_demo.py
```
> Runs default params. For custom runs use `--text`, `--voice`, `--output`, `--speed`,
> `--trace/--no-trace`, `--torch-stft-fallback`, `--torch-phase-fallback`,
> `--pcc-check/--no-pcc-check`. e.g.:
> ```bash
> python models/experimental/kokoro/demo/ttnn_kokoro_full_demo.py \
>   --text "Paris is a beautiful city in the heart of Europe." \
>   --voice af_heart --output out.wav \
>   --torch-stft-fallback --torch-phase-fallback
> ```

## Performance (Blackhole, metal trace, `af_heart`, seed 0)

Warm-replay = steady-state per-chunk latency once the trace is captured; RTF = `latency / audio_s`
(< 1 = faster than real time). Trace replay is **74–301×** faster than the un-traced eager forward.

| phonemes | audio (s) | warm-replay (s) | RTF | cold-capture (s) |
|:--------:|:---------:|:---------------:|:---:|:----------------:|
| 44 | 2.95 | 1.68 | 0.571 | 506.8 |
| 195 | 11.70 | 4.85 | 0.415 | 359.3 |
| 293 | 17.15 | 8.21 | 0.479 | 750.0 |
| 440 | 24.65 | 13.54 | 0.549 | 1525.5 |
| 489 (≈max) | 26.93 | 15.18 | 0.564 | 2348.4 |

> The forward is **dispatch-bound** (device time ≪ wall-clock), so metal tracing is the primary
> latency lever. Cold capture is a one-time per-length cost, amortised by replay.

## Quality / PCC results (`af_heart`, seed 0, deterministic)

All configs PASS every gate (WER < 30%, mel-PCC > 0.95, SECS > 0.95, cFW2VD < 8.0):

| config | WER ↓ | mel PCC ↑ | SECS ↑ | cFW2VD ↓ |
|--------|:-----:|:---------:|:------:|:--------:|
| `stft + phase` (config E) | 0.00% | 0.9928 | 0.9892 | 2.06 |
| `phase` fallback | 0.00% | 0.9929 | 0.9922 | 1.54 |
| `no_fallback` (fully on-device) | 0.00% | 0.9722 | 0.9571 | 3.64 |

Per-module PCC (real weights, full HF config) holds **> 0.999** against the float32 reference.

## More details
See [models/experimental/kokoro/README.md](README.md) for the architecture diagram, module→file map,
dependency versions, the full quality/config matrix, the environment-variable reference, and the
numerical-failure proof tests.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ign/kokoro-latest-extended)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ign/kokoro-latest-extended)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:ign/kokoro-latest-extended)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ign/kokoro-latest-extended)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ign/kokoro-latest-extended)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ign/kokoro-latest-extended)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ign/kokoro-latest-extended)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ign/kokoro-latest-extended)

### CODE REVIEW

File: models/experimental/kokoro/demo/ttnn_kokoro_full_demo.py [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+    """
+    import librosa
+    import numpy as np
+
+    x = audio.detach().float().flatten().cpu().numpy().astype(np.float32)
+    mel = librosa.feature.melspectrogram(y=x, sr=sample_rate, n_fft=1024, hop_length=256, n_mels=80, power=2.0)
+    return torch.from_numpy(librosa.power_to_db(mel, ref=np.max))
+
+
+def main() -> int:
```
Comment (atupe-tt) [inline models/experimental/kokoro/demo/ttnn_kokoro_full_demo.py]: Make sure to follow the pytest convention that we use for model demos
